src/main/java/spark/SparkAPI.java

```diff
@@ -0,0 +1,1132 @@
+package spark;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import spark.exception.ExceptionHandlerImpl;
+import spark.exception.ExceptionMapper;
+import spark.route.HttpMethod;
+import spark.route.RouteMatcherFactory;
+import spark.route.SimpleRouteMatcher;
+import spark.servlet.SparkFilter;
+import spark.utils.SparkUtils;
+import spark.webserver.SparkServer;
+import spark.webserver.SparkServerFactory;
+
+import java.util.concurrent.CountDownLatch;
+
+public class SparkAPI {
+    private static final Logger LOG = LoggerFactory.getLogger(SparkAPI.class);
+
+    public final int SPARK_DEFAULT_PORT = 4567;
+    static final String DEFAULT_ACCEPT_TYPE = "*/*";
+
+    private boolean initialized = false;
+
+    private int port = SPARK_DEFAULT_PORT;
+    private String ipAddress = "0.0.0.0";
+
+    private String keystoreFile;
+    private String keystorePassword;
+    private String truststoreFile;
+    private String truststorePassword;
+
+    private String FileFolder = null;
+    private String externalStaticFileFolder = null;
+
+    private int maxThreads = -1;
+    private int minThreads = -1;
+    private int threadIdleTimeoutMillis = -1;
+
+    private SparkServer server;
+    private SimpleRouteMatcher routeMatcher;
+    private boolean runFromServlet;
+
+    private boolean servletStaticLocationSet;
+    private boolean servletExternalStaticLocationSet;
+
+    private CountDownLatch latch = new CountDownLatch(1);
+
+    /**
+     * Set the IP address that Spark should listen on. If not called the default
+     * address is '0.0.0.0'. This has to be called before any route mapping is
+     * done.
+     *
+     * @param ipAddress The ipAddress
+     */
+    public synchronized void ipAddress(String ipAddress) {
+        if (initialized) {
+            throwBeforeRouteMappingException();
+        }
+        this.ipAddress = ipAddress;
+    }
+
+    /**
+     * Set the port that Spark should listen on. If not called the default port
+     * is 4567. This has to be called before any route mapping is done.
+     * If provided port = 0 then the an arbitrary available port will be used.
+     *
+     * @param port The port number
+     */
+    public synchronized void port(int port) {
+        if (initialized) {
```

Comment:
To my sensibilities, it would be preferable to separate the pieces that have to be initialized before the routes, and the routes themselves. Then, all of this extra complication could be eliminated in favor of an immutable core configuration (the pre-route bits), and and immutable set of filters/routes.  This would remove the necessity of managing the latch, and all the state.